src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -896,6 +918,9 @@ internal async Task Close(bool connectionClosing, bool async, bool isDisposing)
             }
 
             await Cleanup(async, connectionClosing, isDisposing);
+
+            if (consumeException is not null)
```

Comment:
Good solution. One possible micro-optimization would be to move the await Cleanup with this condition into the switch/case above (and so to duplicate the call to Cleanup under the Closed/Broken cases). I'm assuming that users typically consume the reader in user code, and only call Close/Dispose after that - in which case we save the needless condition (since Consume wasn't even called).